Support search indexes in mappings and SchemaManager

jmikola opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | no
| Fixed issues | [PHPORM-38](https://jira.mongodb.org/browse/PHPORM-38)

#### Summary

<!-- Provide a summary your change. -->

jmikola committed:
Model search index mappings

jmikola committed:
Store default search index name in ClassMetadata

jmikola committed:
Support search indexes in SchemaManager

alcaeus commented on doctrine-mongo-mapping.xsd:
```diff
@@ -466,6 +467,96 @@
   </xs:complexType>
 
+  <xs:complexType name="search-indexes">
+    <xs:choice minOccurs="0" maxOccurs="unbounded">
```
Is there a reason to have an empty `search-indexes` element anywhere? Otherwise you could consider using `minOccurs="1"`.

jmikola replied:
This was copied from the `<indexes>` definition, but I see no reason we can't use `minOccurs="1"` here. I'll leave `<indexes>` as-is, though.

The actual change here will just be removing `minOccurs="0"`, since it defaults to one.

alcaeus commented on lib/Doctrine/ODM/MongoDB/SchemaManager.php:
```diff
@@ -318,6 +321,140 @@ public function deleteDocumentIndexes(string $documentName, ?int $maxTimeMs = nu
+    /**
+     * Create search indexes for all mapped document classes.
+     */
+    public function createSearchIndexes(): void
```
Leaving a note to follow up on later, but these methods aren't invoked by the schema commands yet. Since search indexes are an Atlas-only feature, I would suggest duplicating the existing logic for managing indexes instead of combining them with regular indexes.

jmikola replied:
I'm still working through this in the command classes and forgot to preemptively leave a comment about this last week. I was on the fence about creating entirely new methods since that would change the public interface for AbstractCommand. If you're not concerned about that being a BC break, I have no qualms in doing so.

Otherwise, I planned to incorporate this into the `--index` option and introduce a supporting option to enable/disable search indexes. That feels clunkier though, so I'm certainly in favor of a new, explicit option.

alcaeus replied:
While `AbstractCommand` is not marked as internal, I'm not particularly concerned about people extending it, and we can work around it by providing a default implementation for the new methods instead of leaving them empty to avoid compile errors. Any objections to that plan @malarzm?

jmikola replied:
A default implementation seems prudent, even if it'd only be useful internally for ShardCommand (where I'm presently throwing a BadMethodCallException like other unsupported methods).

malarzm replied:
I'd say do whatever you guys need with `AbstractCommand`. I deem it highly unlikely that somebody extended those. And if they did they'll be smart enough to fix it themselves ;)

alcaeus reviewed (commented):
Initial review looks good. There's a number of PHPStan issues due to not properly specifying all array key/value pairs. I'll let you decide whether it's worth specifying them all or whether you'd rather add the errors to the baseline.

jmikola committed:
Support search indexes in CLI commands

Adds default implementations for "process" methods in AbstractCommand, which throw BadMethodCallException.

Renames internal methods in ShardCommand to no longer override "index" base methods, since sharding methods in SchemaManager do much more than process indexes.

jmikola committed:
Require at least one <search-index> within <search-indexes>

jmikola committed:
SchemaManager tests

jmikola committed:
Type information for storedSource and synonym structs

jmikola committed:
phpcbf fixes

jmikola committed:
Update phpstan baseline for unmodeled search index structs

The fields struct is recursive, which is not supported by phpstan. The analyzers struct may be technically possible to model, but the complexity isn't worth the effort.

jmikola committed:
Update psalm baseline for XmlDriver

jmikola committed:
Require driver 1.17+ for search index APIs

jmikola commented:
Sharding failures seem unrelated: https://github.com/doctrine/mongodb-odm/actions/runs/8914210744/job/24481504241?pr=2630

GromNaN commented on lib/Doctrine/ODM/MongoDB/Mapping/Annotations/SearchIndex.php:
```diff
@@ -0,0 +1,39 @@
+     */
+    public function __construct(
+        public ?string $name = null,
+        public ?bool $dynamic = null,
```
Wouldn't it be better to set actual default values rather than nullable types? That would document the default values in the code directly.

```suggestion
        public bool $dynamic = false,
```

But then, my suggestion isn't consistent with the rest of the Doctrine attributes.

jmikola replied:
I did originally consider this, but thought it best to minimize what ends up in the definition shipped to `createSearchIndexes`.

alcaeus replied:
Agree - we've typically used nullable bools when we want to omit values from the command unless they were explicitly specified, and the only way to achieve this is by using a `null` default value.

GromNaN commented on tests/Documents/CmsArticle.php:
```diff
@@ -6,9 +6,9 @@
 
-#[Index(keys: ['topic' => 'asc'])]
+#[ODM\Index(keys: ['topic' => 'asc'])]
+#[ODM\SearchIndex(dynamic: true)]
```
Could you use all the fields in the attribute to make the test more exhaustive?

jmikola replied:
The mapping tests are exhaustive in their use of all options. This model is only used for the SchemeManager tests, in which we assert that _something_ is passed along to the mocked MongoDB\Collection methods. That is based on prior art for the standard index tests.

alcaeus commented on lib/Doctrine/ODM/MongoDB/Tools/Console/Command/Schema/AbstractCommand.php:
```diff
@@ -34,23 +36,81 @@ protected function configure()
+     *
+     * @throws BadMethodCallException
+     */
+    protected function processDocumentCollection(SchemaManager $sm, string $document, ?int $maxTimeMs, ?WriteConcern $writeConcern)
+    {
+        throw new BadMethodCallException('This command does not support collections');
+    }
```
👍 for adding default implementations for all methods in this class.

alcaeus commented on lib/Doctrine/ODM/MongoDB/Tools/Console/Command/Schema/CreateCommand.php:
```diff
@@ -53,17 +60,23 @@ private function doExecute(InputInterface $input, OutputInterface $output): int
+                self::COLLECTION => 'Collection',
+                self::INDEX => 'Index',
+                self::SEARCH_INDEX => 'SearchIndex',
+                default => throw new UnhandledMatchError($option),
```
This is standard behaviour in PHP, see https://3v4l.org/LU66Z.
```suggestion
```

jmikola replied:
Indeed it is; however, PHPStan doesn't approve: https://github.com/phpstan/phpstan/issues/4451

I can replace this with a baseline addition. I'm currently only relying on the baseline for the truly unreasonable issues (`fields` and `analyzers` options).

alcaeus replied:
I'll leave this up to you and chalk this up as another point against PHPStan. Feel free to resolve after reading.

jmikola replied:
In the interest of keeping the code cleaner, I'll remove the useless `default` cases and track this in the PHPStan baseline.

alcaeus commented on lib/Doctrine/ODM/MongoDB/Tools/Console/Command/Schema/DropCommand.php:
```diff
@@ -50,17 +59,24 @@ private function doExecute(InputInterface $input, OutputInterface $output): int
+                self::COLLECTION => 'Collection',
+                self::INDEX => 'Index',
+                self::SEARCH_INDEX => 'SearchIndex',
+                default => throw new UnhandledMatchError($option),
```
```suggestion
```

alcaeus commented on lib/Doctrine/ODM/MongoDB/Tools/Console/Command/Schema/UpdateCommand.php:
```diff
@@ -28,14 +27,16 @@ protected function configure()
         $this
             ->setName('odm:schema:update')
             ->addOption('class', 'c', InputOption::VALUE_OPTIONAL, 'Document class to process (default: all classes)')
+            ->addOption('disable-search-indexes', null, InputOption::VALUE_NONE, 'Do not update search indexes')
```
I'm aware the naming is consistent with `disable-validators`, but wouldn't `skip-search-indexes` be more appropriate?

jmikola replied:
SGTM. I'll make the change.

Per our earlier discussion about default behavior for CreateCommand and DropCommand, do you think a similar option is warranted for those? Since search indexes are included in the default `$createOrder` and `$dropOrder` lists, I suppose users would need to tweak the CLI args either way (be it `--skip-search-indexes` or explicitly specifying the things to process). So I'm not convinced it's needed there, in contrast to UpdateCommand.

alcaeus replied:
I'd say for now we can leave create and drop as-is, but I agree that it's unfortunate to have to redefine the entire list when wanting to exclude a single item, especially as that can create issues when we add new functionality. A dedicated `skip-<feature>` or `include-<feature>` (for optional things) might be better, despite it meaning more command-line arguments. Let's create a ticket for this and put it on the map for a future refactoring of those commands.

jmikola replied:
Opened https://github.com/doctrine/mongodb-odm/issues/2634. I only proposed `skip-<feature>` as the commands already support opt-in behavior. For example, CreateCommand's `--collection` option would accomplish the same thing as `include-collection`, assuming both would disable the `$createOrder` and expect everything to be explicitly requested (as `--collection` does today).

jmikola replied:
Renamed this to `skip-search-indexes` and added the same option to CreateCommand and DropCommand. #2634 can be used for renaming `disable-validators` and adding additional `skip-<feature>` options if desired.

alcaeus replied:
Sounds good, thank you!

jmikola committed:
Add skip-search-indexes option to schema CLI commands

Currently, commands can either process all definitions (default behavior) or specify individual definitions. This allows the commands to rely on default behavior (e.g. $createOrder) but omit processing of search indexes, which may be more stringent requirements.

Note: this is similar to the disable-validators option that already existed in UpdateCommand; however, #2634 suggests renaming that if additional "skip" options are introduced.

jmikola commented:
Psalm is failing due to [two unused baseline entries](https://github.com/doctrine/mongodb-odm/actions/runs/8972457667/job/24640499243?pr=2630#step:6:27), but I'm unable to reproduce that locally with PHP 8.2 (same as is used in CI):

```
Error: lib/Doctrine/ODM/MongoDB/UnitOfWork.php:0:0: UnusedBaselineEntry: Baseline for issue "TypeDoesNotContainType" has 1 extra entry. (see https://psalm.dev/316)
Error: tests/Doctrine/ODM/MongoDB/Tests/SchemaManagerTest.php:0:0: UnusedBaselineEntry: Baseline for issue "TypeDoesNotContainType" has 1 extra entry. (see https://psalm.dev/316)
```

UnitOfWork is also entirely unrelated to this PR, so I'm not sure what to make of that.

alcaeus commented:
> Psalm is failing due to [two unused baseline entries](https://github.com/doctrine/mongodb-odm/actions/runs/8972457667/job/24640499243?pr=2630#step:6:27), but I'm unable to reproduce that locally with PHP 8.2 (same as is used in CI):

This could be related to using an outdated psalm version - I was able to reproduce it locally with psalm 5.24.0.

alcaeus reviewed (approved):
LGTM with psalm issues fixed.

jmikola committed:
Update baseline for Psalm 5.24.0

jmikola commented:
Docs are still in progress (very not fun).

jmikola committed:
SearchIndex annotation docs

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/Mapping/AbstractMappingDriverTestCase.php:
```diff
@@ -634,6 +686,27 @@ public function testEnumType(): void
+ *       },
+ *       {"type"="autocomplete"},
+ *     },
+ *     "embedded_phone_number"={"type"="embeddedDocuments", "dynamic"=true},
```
Note: [embeddedDocuments](https://www.mongodb.com/docs/atlas/atlas-search/field-types/embedded-documents-type/) is used because the field is mapped as EmbedMany:

```
@ODM\EmbedMany(targetDocument=Phonenumber::class, name="embedded_phone_number")
```

If the relationship was EmbedOne then [document](https://www.mongodb.com/docs/atlas/atlas-search/field-types/document-type/) would be appropriate.

jmikola replied:
I've ensured one of the docs examples uses `embeddedDocuments`.

jmikola committed:
Search indexes chapter

alcaeus reviewed (approved):
Gave the docs a read and they look good to me. Since we can't easily test docs generation, I'd suggest we merge this as is and fix any RST issues after the fact.

stof commented on docs/en/reference/search-indexes.rst:
```diff
@@ -0,0 +1,188 @@
+
+        /**
+         * @Document
+         * @SearchIndex(
```
shouldn't this show an eample using attributes rather than annotations ?

jmikola replied:
It looks like that was addressed in https://github.com/doctrine/mongodb-odm/commit/37013acdf1db8c10e2982e0700f72682d1fd2339
